File: 0144-binary-tree-preorder-traversal/solution.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
void preorder(struct TreeNode *node,int *,int *);
int count(struct TreeNode *node);
int* preorderTraversal(struct TreeNode* root, int* returnSize) {
    if(root==NULL)
    {
        *returnSize=0;
        return NULL;
    }
    int s=count(root);
    int *ans=(int *)malloc(s*sizeof(int));
    int i=0;
    preorder(root,ans,&i);
    *returnSize=s;
    return ans;
}
int count(struct TreeNode *node)
{
    if(node==NULL)
    return 0;
    return 1+count(node->left)+count(node->right);
}
void preorder(struct TreeNode *node,int *ans,int *i)
{
    if(node==NULL)
    return;
    ans[(*i)++]=node->val;
    preorder(node->left,ans,i);
    preorder(node->right,ans,i);
}

File: 0144-binary-tree-preorder-traversal/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> l=new ArrayList<Integer>();
        root=preorder(root,l);
        return l;
    }
    static TreeNode preorder(TreeNode node,List<Integer> l)
    {
        if(node==null)
        return null;
        l.add(node.val);
        preorder(node.left,l);
        preorder(node.right,l);
        return node;
    }
}
